crates/sui-storage/src/event_store/sql.rs

```diff
@@ -0,0 +1,513 @@
+        // TODO: benchmark
+        // TODO: use techniques in https://docs.rs/sqlx-core/0.5.13/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values
+        // to execute all inserts in a single statement?
+        // TODO: See https://kerkour.com/high-performance-rust-with-sqlite
+        for event in events {
+            // If batching, turn off persistent to avoid caching as we may fill up the prepared statement cache
+            let insert_tx_q = sqlx::query(SQL_INSERT_TX).persistent(true);
+            let module_id = event.event.module_id();
+            let event_type = EventType::from(&event.event);
+            // TODO: use batched API?
+            insert_tx_q
+                .bind(event.timestamp as i64)
+                .bind(checkpoint_num as i64)
+                .bind(event.tx_digest.map(|txd| txd.to_bytes()))
+                .bind(event_type as u16)
+                .bind(module_id.clone().map(|mid| mid.address().to_vec()))
+                .bind(module_id.map(|mid| mid.name().to_string()))
+                .bind(event.event.object_id().map(|id| id.to_vec()))
+                .bind(event_to_json(event))
+                .execute(&self.pool)
+                .await?;
+        }
+        Ok(())
+    }
+
+    async fn events_for_transaction(
+        &self,
+        digest: TransactionDigest,
+    ) -> Result<Self::EventIt, EventStoreError> {
+        let rows = sqlx::query(TX_QUERY)
+            .persistent(true)
+            .bind(digest.to_bytes())
+            .map(sql_row_to_event)
+            .fetch_all(&self.pool)
+            .await?;
+        Ok(rows.into_iter())
+    }
+
+    async fn events_by_type(
+        &self,
+        start_time: u64,
+        end_time: u64,
+        event_type: EventType,
+        limit: usize,
+    ) -> Result<Self::EventIt, EventStoreError> {
+        check_limit(limit)?;
+        let rows = sqlx::query(QUERY_BY_TYPE)
+            .persistent(true)
+            .bind(start_time as i64)
+            .bind(end_time as i64)
+            .bind(event_type as u16)
+            .bind(limit as i64)
+            .map(sql_row_to_event)
+            .fetch_all(&self.pool)
+            .await?;
+        Ok(rows.into_iter())
+    }
+
+    async fn event_iterator(
+        &self,
+        start_time: u64,
+        end_time: u64,
+        limit: usize,
+    ) -> Result<Self::EventIt, EventStoreError> {
+        check_limit(limit)?;
+        let rows = sqlx::query(TS_QUERY)
+            .bind(start_time as i64)
+            .bind(end_time as i64)
+            .bind(limit as i64)
+            .map(sql_row_to_event)
+            .fetch_all(&self.pool)
+            .await?;
+        Ok(rows.into_iter())
+    }
+
+    async fn events_by_checkpoint(
+        &self,
+        start_checkpoint: u64,
+        end_checkpoint: u64,
+        limit: usize,
+    ) -> Result<Self::EventIt, EventStoreError> {
+        // TODO: a limit maybe doesn't make sense here.  May change to unbounded iterator?
```

Comment:
I suppose the limit is to avoid draining a node's resources to get a tons of things out of disk? Not sure what does it mean by `unbounded iterator`.

But I agree with you that it's not logic complete, if a checkpoint has many events, then a client has no way to get the events on the later part. Maybe pagination is what we are looking for ... Make a TODO?

Reply:
It is already a TODO, I think I'd prefer to implement and use the streaming method (async streams) instead of pagination, but we can def discuss later.